extensions-contrib/opencensus-extensions/src/main/java/org/apache/druid/data/input/opencensus/protobuf/OpenCensusProtobufReader.java

```diff
@@ -81,29 +97,75 @@ public OpenCensusProtobufReader(
   }
 
   @Override
-  public CloseableIterator<InputRow> read()
+  public CloseableIterator<InputRow> read() throws IOException
   {
-    Supplier<Iterator<InputRow>> supplier = Suppliers.memoize(() -> readAsList().iterator());
-    return CloseableIterators.withEmptyBaggage(new Iterator<InputRow>() {
-      @Override
-      public boolean hasNext()
-      {
-        return supplier.get().hasNext();
-      }
-      @Override
-      public InputRow next()
-      {
-        return supplier.get().next();
+    return newReader(whichReader()).read();
+  }
+
+  class OpenCensusProtobufReaderInternal implements InputEntityReader
+  {
+    @Override
+    public CloseableIterator<InputRow> read()
+    {
+      return CloseableIterators.withEmptyBaggage(readAsList().iterator());
+    }
+
+    @Override
+    public CloseableIterator<InputRowListPlusRawValues> sample() throws IOException
+    {
+      return OpenCensusProtobufReader.this.sample();
+    }
+  }
+
+  public InputEntityReader newReader(ProtobufReader which)
+  {
+    switch(which) {
+      case OPEN_TELEMETRY:
+        return new OpenTelemetryMetricsProtobufReader(
+            dimensionsSpec,
+            source,
+            metricDimension,
+            valueDimension,
+            metricLabelPrefix,
+            resourceLabelPrefix
+        );
+      case OPEN_CENSUS:
+      default:
+        return new OpenCensusProtobufReaderInternal();
+    }
+  }
+
+  public ProtobufReader whichReader()
+  {
+    // assume InputEntity is always defined in a single classloader (the kafka-indexing-service classloader)
+    // so we only have to look it up once. To be completely correct we should cache the method based on classloader
+    MethodHandle getHeaderMethod = KafkaUtils.lookupGetHeaderMethod(
+        source.getEntity().getClass().getClassLoader(),
+        VERSION_HEADER_KEY
+    );
+
+    try {
+      byte[] versionHeader = (byte[]) getHeaderMethod.invoke(source.getEntity());
+      if (versionHeader != null) {
+        int version =
+            ByteBuffer.wrap(versionHeader).order(ByteOrder.LITTLE_ENDIAN).getInt();
+        if (version == OPENTELEMETRY_FORMAT_VERSION) {
+          return OPEN_TELEMETRY;
+        }
       }
-    });
+    }
+    catch (Throwable t) {
+      // assume input is opencensus if something went wrong
+    }
+    return OPEN_CENSUS;
   }
 
   List<InputRow> readAsList()
   {
     try {
-      return parseMetric(Metric.parseFrom(source.getBuffer()));
+      return parseMetric(Metric.parseFrom(source.open()));
     }
-    catch (InvalidProtocolBufferException e) {
+    catch (IOException e) {
```

Comment:
using the suggestion above we can also avoid having to catch additional IOExceptions

Reply:
looks like we revert this change now and only catch InvalidProtocolBufferException?

Reply:
I reinstated `InvalidProtocolBufferException` but that doesn't remove the IOException in `sample` and `read` methods.